src/object/cli_obj.c

```diff
@@ -2627,3 +2673,540 @@ dc_obj_query_key(tse_task_t *api_task)
+				if (arg_s->buf == NULL)
+					D_GOTO(out, rc = -DER_NOMEM);
+			}
+
+			rc = daos_enum_dkeys_do_list(oh, &epoch, arg_s, list_cb,
+						     DIOF_TO_SPEC_SHARD);
+
+			if (arg_s->need_retry) {
+				arg_s->need_retry = 0;
+				arg_s->has_retried = 1;
+			}
+
+			/* We used stale pool map, need to check new leader. */
+			if (rc == -DER_STALE) {
+				D_DEBUG(DB_TRACE, "(reps %d, start %d) Got "
+					"stale when list from non-leader (%d), "
+					"retry (with new leader)\n",
+					reps, start, i + start);
+				goto leader;
+			}
+
+			/* If non-leader has more things than leader, then:
+			 * 1. The leader and non-leader may be inconsistent. Or
+			 * 2. There are some modifications committed after list
+			 *    from the leader.
+			 *
+			 * We need to re-list the dkeys with leader. If leader
+			 * has more things, it may be the 2nd case. Otherwise,
+			 * it is the 1st case.
+			 */
+			if (rc > 0) {
+				if (dc_obj_need_recheck(oeda, reps, start,
+							leader, i, flags,
+							"re-list_1"))
+					goto list;
+
+				D_INFO(DF_OID" (reps %d, inconsistent) shard "
+				       "%d (l) has less data than shard %d\n",
+				       DP_OID(oid), reps,
+				       leader + start, i + start);
+				D_GOTO(out, rc = -DER_MISMATCH);
+			}
+
+			if (rc < 0) {
+				if (rc != -DER_NONEXIST)
+					D_GOTO(out, rc);
+
+				rc = 0;
+				arg_s->lost_shard = 1;
+			} else {
+				arg_s->lost_shard = 0;
+			}
+
+			if (arg_l->lost_shard) {
+				if (arg_s->lost_shard)
+					continue;
+
+				D_INFO(DF_OID" (reps %d, inconsistent) shard "
+				       "%d (l) does not exists, but shard %d "
+				       "does exists.\n", DP_OID(oid), reps,
+				       leader + start, i + start);
+				D_GOTO(out, rc = -DER_MISMATCH);
+			}
+
+			if (arg_s->lost_shard) {
+				/* Non-leader does not exist, but leader seems
+				 * there. It may because someone destroyed the
+				 * object after we list from leader. Let's try
+				 * to re-list from leader.
+				 */
+				if (!arg_l->has_retried &&
+				    dc_obj_need_recheck(oeda, reps, start,
+							leader, i, flags,
+							"re-list_2"))
+					goto list;
+
+				D_INFO(DF_OID" (reps %d, inconsistent) shard "
+				       "%d (l) does exists, but shard %d does "
+				       "not exists.\n", DP_OID(oid), reps,
+				       leader + start, i + start);
+				D_GOTO(out, rc = -DER_MISMATCH);
+			}
+
+			if (arg_l->size != arg_s->size ||
+			    memcmp(arg_l->buf, arg_s->buf, arg_l->size) != 0) {
+				/* Non-leader has different keys. It may because
+				 * someone changed the object after we list from
+				 * leader. Let's re-list from leader.
+				 */
+				if (!arg_l->has_retried &&
+				    dc_obj_need_recheck(oeda, reps, start,
+							leader, i, flags,
+							"re-list_3"))
+					goto list;
+
+				D_INFO(DF_OID" (reps %d, inconsistent) shard "
+				       "%d (l) and shard %d have inconsistent "
+				       "keys/recs list\n", DP_OID(oid), reps,
+				       leader + start, i + start);
+				D_GOTO(out, rc = -DER_MISMATCH);
+			}
```

Comment:
The issue is that from the client perspective, it does NOT know whether there is concurrent IO or not. It has to retry with server to check whether the "inconsistency" is real inconsistent or just caused by some concurrent IO. So after the retry, if the client found that there is concurrent IO, then it can either 1) ignore the difference and go ahead (current patch does) or 2) return -EBUSY as you suggested. So if without the retry logic, directly return -EBUSY is wrong.

Reply:
I mean if the client find there are uncommitted record during verification, then it can return EBUSY. I think client can know this for iteration and fetch, why through retry?

Reply:
I cannot follow your idea. We do the iterations on each replica one by one, how can we guarantee there is no race modification among these iterations on different replicas?

For example: when the client does iteration on the replica_A, there is some DTX in local 'prepared' status, but waiting for other replicas; when we move to iterate replica_B, such DTX is committed. Then iteration from replica_B will get more things than from replica_A. That is just one possible case. Under such case, if we re-iterate the replica_A, we will find replica_A also contains more things than the former iteration, so we can know the former 'inconsistent' between replica_A and replica_B is fake, caused by concurrent modification. Then you can return -EBUSY if you do not want to touch the object in medication, or can handle as my current patches does. Anyway, if without the re-iteration (retry), the client cannot make the right decision.

Reply:
More special case is that when we iterate replica_A, the DTX_m is not started yet, but when we moved to iterate replica_B, the DTX_m happened and committed. Such case can NOT be handled via your -EBUSY way.

Reply:
In my current implementation, the 'retry' logic will be bypass if we call the API daos_obj_verify() with 'DOVF_IMMUTABLE' flag. It is the caller duty to guarantee that there is no concurrent modification during the verification. But from the long perspective, 'retry' logic is necessary and useful. That is why I bypass them via some flag instead of totally removing.

Reply:
When we verify the object, it should exclusively connect the pool or open the container, i.e. make sure no one access the object at the same time, though this is not implemented yet.

As I understand, we do not intend to implement a on-line "fsck" here, just a verification tool.

Reply:
You set too much restriction about the condition of using the consistency verification tool, such as "exclusively connect the pool or open the container", as to if I am the DAOS user, I will NOT use it because it is not easy to match these requirements. If I am the user, what I expect is just a API, I do not care and I should not know whether there is concurrent modification or not, it is the API's internal logic handle related rare conditions.

The current implementation is not a on-link fsck, we are far away from such purpose. But we can make it as the base. I do not want to re-implement another similar tool if we want to support on-link fsck in the future.

Reply:
I tend to agree that this complicated 'retry on race' is not necessary and flawed.

I think the verify procedure should only be performed on a confined epoch range ([0, epoch_A], for instance), and it's easy to guarantee that no updates will happen under the upper bound of epoch, so it's not necessary to introduce this 'retry on race' logic to handle the race of verify vs. update.

What we need to take care is the concurrent snapshot deletion or rebuild, and it's easy to make them exclusive with verify in upper layer.

Please keep in mind that that 'verify' must be done in 'logical view' manner (verify that replica1, replica2 has same logical view from certain specified epoch). Because of the VOS aggregation running in background, the physical enumerations on different replicas are likely to be different.

Reply:
> I tend to agree that this complicated 'retry on race' is not necessary and flawed.
> 

What is the 'flawed'?

> I think the verify procedure should only be performed on a confined epoch range ([0, epoch_A], for instance)

From the DAOS user view, it may does not know the epoch_A, so for most of cases, he/she may just specifies [0, MAX] as the range. That will be the same case as current use case.

> and it's easy to guarantee that no updates will happen under the upper bound of epoch, so it's not necessary to introduce this 'retry on race' logic to handle the race of verify vs. update.

How can you guarantee for that? The modification against different keys can go in parallel, when epoch_A is committed, the older one epoch_B may be not committed yet.

The "retry" logic in the patch is used to check the modification with non-committable DTX when we get list epoch on the leader replica. We do not care about other subsequent modifications, because their epoch will be newer than the list epoch.


> Please keep in mind that that 'verify' must be done in 'logical view' manner (verify that replica1, replica2 has same logical view from certain specified epoch). Because of the VOS aggregation running in background, the physical enumerations on different replicas are likely to be different.

Would you please to explain that in some detail? Assume the original replicas are consistent. At some time point, vos aggregation is triggered, does the vos aggregation on different replicas will generate different layout? I mean after the aggregation done. Can we detect whether vos aggregation is running or not? if we can, then we can ask the verification logic to return -BUSY if vos aggregation is running.